#+STARTUP: indent
#+STARTUP: content
* MDA map2.sh pipeline
Note 
bwa merge -r infers read group from file name
#+BEGIN_SRC sh
cd /projects/flowcell61_A_paired_BCX
echo align the reads at `date`
echo `pwd`
echo `date` map Cap474-101069-23_GAGTGG_L004_R1_001.fastq.gz..
/projects/tools/bwa-0.6.2/bwa  aln -t 1 /projects/database/reference/hs37d5.fa data/IPCT-BCX-PT155-101068-A-TU-A01/gz/Cap474-101069-23_GAGTGG_L004_R1_001.fastq.gz>data/IPCT-BCX-PT155-101068-A-TU-A01/bam/Cap474-101069-23_GAGTGG_L004_R1_001.fastq.gz.sai
echo `date` map Cap474-101069-23_GAGTGG_L004_R1_001.fastq.gz done .

echo `date` map Cap474-101069-23_GAGTGG_L004_R1_002.fastq.gz..
/projects/tools/bwa-0.6.2/bwa  aln -t 1 /projects/database/reference/hs37d5.fa data/IPCT-BCX-PT155-101068-A-TU-A01/gz/Cap474-101069-23_GAGTGG_L004_R1_002.fastq.gz>data/IPCT-BCX-PT155-101068-A-TU-A01/bam/Cap474-101069-23_GAGTGG_L004_R1_002.fastq.gz.sai
echo `date` map Cap474-101069-23_GAGTGG_L004_R1_002.fastq.gz done .

echo `date` map Cap474-101069-23_GAGTGG_L004_R1_003.fastq.gz..
/projects/tools/bwa-0.6.2/bwa  aln -t 1 /projects/database/reference/hs37d5.fa data/IPCT-BCX-PT155-101068-A-TU-A01/gz/Cap474-101069-23_GAGTGG_L004_R1_003.fastq.gz>data/IPCT-BCX-PT155-101068-A-TU-A01/bam/Cap474-101069-23_GAGTGG_L004_R1_003.fastq.gz.sai
echo `date` map Cap474-101069-23_GAGTGG_L004_R1_003.fastq.gz done .

echo `date` map Cap474-101069-23_GAGTGG_L004_R1_004.fastq.gz..
/projects/tools/bwa-0.6.2/bwa  aln -t 1 /projects/database/reference/hs37d5.fa data/IPCT-BCX-PT155-101068-A-TU-A01/gz/Cap474-101069-23_GAGTGG_L004_R1_004.fastq.gz>data/IPCT-BCX-PT155-101068-A-TU-A01/bam/Cap474-101069-23_GAGTGG_L004_R1_004.fastq.gz.sai
echo `date` map Cap474-101069-23_GAGTGG_L004_R1_004.fastq.gz done .

echo `date` map Cap474-101069-23_GAGTGG_L004_R2_001.fastq.gz..
/projects/tools/bwa-0.6.2/bwa  aln -t 1 /projects/database/reference/hs37d5.fa data/IPCT-BCX-PT155-101068-A-TU-A01/gz/Cap474-101069-23_GAGTGG_L004_R2_001.fastq.gz>data/IPCT-BCX-PT155-101068-A-TU-A01/bam/Cap474-101069-23_GAGTGG_L004_R2_001.fastq.gz.sai
echo `date` map Cap474-101069-23_GAGTGG_L004_R2_001.fastq.gz done .

echo `date` map Cap474-101069-23_GAGTGG_L004_R2_002.fastq.gz..
/projects/tools/bwa-0.6.2/bwa  aln -t 1 /projects/database/reference/hs37d5.fa data/IPCT-BCX-PT155-101068-A-TU-A01/gz/Cap474-101069-23_GAGTGG_L004_R2_002.fastq.gz>data/IPCT-BCX-PT155-101068-A-TU-A01/bam/Cap474-101069-23_GAGTGG_L004_R2_002.fastq.gz.sai
echo `date` map Cap474-101069-23_GAGTGG_L004_R2_002.fastq.gz done .

echo `date` map Cap474-101069-23_GAGTGG_L004_R2_003.fastq.gz..
/projects/tools/bwa-0.6.2/bwa  aln -t 1 /projects/database/reference/hs37d5.fa data/IPCT-BCX-PT155-101068-A-TU-A01/gz/Cap474-101069-23_GAGTGG_L004_R2_003.fastq.gz>data/IPCT-BCX-PT155-101068-A-TU-A01/bam/Cap474-101069-23_GAGTGG_L004_R2_003.fastq.gz.sai
echo `date` map Cap474-101069-23_GAGTGG_L004_R2_003.fastq.gz done .

echo `date` map Cap474-101069-23_GAGTGG_L004_R2_004.fastq.gz..
/projects/tools/bwa-0.6.2/bwa  aln -t 1 /projects/database/reference/hs37d5.fa data/IPCT-BCX-PT155-101068-A-TU-A01/gz/Cap474-101069-23_GAGTGG_L004_R2_004.fastq.gz>data/IPCT-BCX-PT155-101068-A-TU-A01/bam/Cap474-101069-23_GAGTGG_L004_R2_004.fastq.gz.sai
echo `date` map Cap474-101069-23_GAGTGG_L004_R2_004.fastq.gz done .

echo `date` sampe ..
/projects/tools/bwa-0.6.2/bwa  sampe /projects/database/reference/hs37d5.fa data/IPCT-BCX-PT155-101068-A-TU-A01/bam/Cap474-101069-23_GAGTGG_L004_R1_001.fastq.gz.sai data/IPCT-BCX-PT155-101068-A-TU-A01/bam/Cap474-101069-23_GAGTGG_L004_R2_001.fastq.gz.sai data/IPCT-BCX-PT155-101068-A-TU-A01/gz/Cap474-101069-23_GAGTGG_L004_R1_001.fastq.gz data/IPCT-BCX-PT155-101068-A-TU-A01/gz/Cap474-101069-23_GAGTGG_L004_R2_001.fastq.gz| /projects/tools/samtools-0.1.19/samtools view  -b -S -|/projects/tools/samtools-0.1.19/samtools sort - data/IPCT-BCX-PT155-101068-A-TU-A01/bam/IPCT-BCX-PT155-101068-A-TU-A01_1

/projects/tools/bwa-0.6.2/bwa  sampe /projects/database/reference/hs37d5.fa data/IPCT-BCX-PT155-101068-A-TU-A01/bam/Cap474-101069-23_GAGTGG_L004_R1_002.fastq.gz.sai data/IPCT-BCX-PT155-101068-A-TU-A01/bam/Cap474-101069-23_GAGTGG_L004_R2_002.fastq.gz.sai data/IPCT-BCX-PT155-101068-A-TU-A01/gz/Cap474-101069-23_GAGTGG_L004_R1_002.fastq.gz data/IPCT-BCX-PT155-101068-A-TU-A01/gz/Cap474-101069-23_GAGTGG_L004_R2_002.fastq.gz| /projects/tools/samtools-0.1.19/samtools view  -b -S -|/projects/tools/samtools-0.1.19/samtools sort - data/IPCT-BCX-PT155-101068-A-TU-A01/bam/IPCT-BCX-PT155-101068-A-TU-A01_2

/projects/tools/bwa-0.6.2/bwa  sampe /projects/database/reference/hs37d5.fa data/IPCT-BCX-PT155-101068-A-TU-A01/bam/Cap474-101069-23_GAGTGG_L004_R1_003.fastq.gz.sai data/IPCT-BCX-PT155-101068-A-TU-A01/bam/Cap474-101069-23_GAGTGG_L004_R2_003.fastq.gz.sai data/IPCT-BCX-PT155-101068-A-TU-A01/gz/Cap474-101069-23_GAGTGG_L004_R1_003.fastq.gz data/IPCT-BCX-PT155-101068-A-TU-A01/gz/Cap474-101069-23_GAGTGG_L004_R2_003.fastq.gz| /projects/tools/samtools-0.1.19/samtools view  -b -S -|/projects/tools/samtools-0.1.19/samtools sort - data/IPCT-BCX-PT155-101068-A-TU-A01/bam/IPCT-BCX-PT155-101068-A-TU-A01_3

/projects/tools/bwa-0.6.2/bwa  sampe /projects/database/reference/hs37d5.fa data/IPCT-BCX-PT155-101068-A-TU-A01/bam/Cap474-101069-23_GAGTGG_L004_R1_004.fastq.gz.sai data/IPCT-BCX-PT155-101068-A-TU-A01/bam/Cap474-101069-23_GAGTGG_L004_R2_004.fastq.gz.sai data/IPCT-BCX-PT155-101068-A-TU-A01/gz/Cap474-101069-23_GAGTGG_L004_R1_004.fastq.gz data/IPCT-BCX-PT155-101068-A-TU-A01/gz/Cap474-101069-23_GAGTGG_L004_R2_004.fastq.gz| /projects/tools/samtools-0.1.19/samtools view  -b -S -|/projects/tools/samtools-0.1.19/samtools sort - data/IPCT-BCX-PT155-101068-A-TU-A01/bam/IPCT-BCX-PT155-101068-A-TU-A01_4

echo `date` sampe done. 

/projects/tools/samtools-0.1.19/samtools view -H data/IPCT-BCX-PT155-101068-A-TU-A01/bam/IPCT-BCX-PT155-101068-A-TU-A01_1.bam>data/IPCT-BCX-PT155-101068-A-TU-A01/bam/header.txt
echo -e "@RG	ID:IPCT-BCX-PT155-101068-A-TU-A01_1	LB:IPCT-BCX-PT155-101068-A-TU-A01	SM:IPCT-BCX-PT155-101068-A-TU-A01">>data/IPCT-BCX-PT155-101068-A-TU-A01/bam/header.txt
echo -e "@RG	ID:IPCT-BCX-PT155-101068-A-TU-A01_2	LB:IPCT-BCX-PT155-101068-A-TU-A01	SM:IPCT-BCX-PT155-101068-A-TU-A01">>data/IPCT-BCX-PT155-101068-A-TU-A01/bam/header.txt
echo -e "@RG	ID:IPCT-BCX-PT155-101068-A-TU-A01_3	LB:IPCT-BCX-PT155-101068-A-TU-A01	SM:IPCT-BCX-PT155-101068-A-TU-A01">>data/IPCT-BCX-PT155-101068-A-TU-A01/bam/header.txt
echo -e "@RG	ID:IPCT-BCX-PT155-101068-A-TU-A01_4	LB:IPCT-BCX-PT155-101068-A-TU-A01	SM:IPCT-BCX-PT155-101068-A-TU-A01">>data/IPCT-BCX-PT155-101068-A-TU-A01/bam/header.txt
echo `date` merge the small bam files..
/projects/tools/samtools-0.1.19/samtools merge -r -h data/IPCT-BCX-PT155-101068-A-TU-A01/bam/header.txt -f data/IPCT-BCX-PT155-101068-A-TU-A01/bam/IPCT-BCX-PT155-101068-A-TU-A01.sorted.bam data/IPCT-BCX-PT155-101068-A-TU-A01/bam/IPCT-BCX-PT155-101068-A-TU-A01_*.bam
echo `date` small bams merged. Sorted.bam obtained.

echo `date` indexing data/IPCT-BCX-PT155-101068-A-TU-A01/bam/IPCT-BCX-PT155-101068-A-TU-A01.sorted.bam..
/projects/tools/samtools-0.1.19/samtools index data/IPCT-BCX-PT155-101068-A-TU-A01/bam/IPCT-BCX-PT155-101068-A-TU-A01.sorted.bam
echo `date` index done.

echo `date` run flagstat..
/projects/tools/samtools-0.1.19/samtools flagstat data/IPCT-BCX-PT155-101068-A-TU-A01/bam/IPCT-BCX-PT155-101068-A-TU-A01.sorted.bam>data/IPCT-BCX-PT155-101068-A-TU-A01/bam/IPCT-BCX-PT155-101068-A-TU-A01.sorted.bam.flagstat
echo `date` flagstat done.
echo `date` MarkDuplicates..
/usr/bin/java -jar /projects/tools/picard-tools-1.80/MarkDuplicates.jar I=data/IPCT-BCX-PT155-101068-A-TU-A01/bam/IPCT-BCX-PT155-101068-A-TU-A01.sorted.bam O=data/IPCT-BCX-PT155-101068-A-TU-A01/bam/IPCT-BCX-PT155-101068-A-TU-A01.rmdup.bam M=data/IPCT-BCX-PT155-101068-A-TU-A01/bam/IPCT-BCX-PT155-101068-A-TU-A01.markdup.metrics AS=true REMOVE_DUPLICATES=true VALIDATION_STRINGENCY=LENIENT
echo `date`  MarkDuplicate Done.
echo remove sorted.bam file..
rm data/IPCT-BCX-PT155-101068-A-TU-A01/bam/IPCT-BCX-PT155-101068-A-TU-A01.sorted.bam
echo sorted.bam deleted.

echo `date` indexing the rmduped files..
/projects/tools/samtools-0.1.19/samtools index data/IPCT-BCX-PT155-101068-A-TU-A01/bam/IPCT-BCX-PT155-101068-A-TU-A01.rmdup.bam
echo `date` index done.

echo `date` mpilup the rmduped bam..
/projects/tools/samtools-0.1.19/samtools mpileup -q 20 -C 50 -f /projects/database/reference/hs37d5.fa data/IPCT-BCX-PT155-101068-A-TU-A01/bam/IPCT-BCX-PT155-101068-A-TU-A01.rmdup.bam | awk '{if($4!=0) print $0}'>data/IPCT-BCX-PT155-101068-A-TU-A01/bam/IPCT-BCX-PT155-101068-A-TU-A01.rmdup.mpileup
echo `date` mpileup done at `date`.

echo `date` running flagstat for rmduped bam..
/projects/tools/samtools-0.1.19/samtools flagstat data/IPCT-BCX-PT155-101068-A-TU-A01/bam/IPCT-BCX-PT155-101068-A-TU-A01.rmdup.bam>data/IPCT-BCX-PT155-101068-A-TU-A01/bam/IPCT-BCX-PT155-101068-A-TU-A01.rmdup.bam.flagstat
echo `date` flagstat done.

rm data/IPCT-BCX-PT155-101068-A-TU-A01/bam/IPCT-BCX-PT155-101068-A-TU-A01_*.bam
echo removing the intermediate bam files..
echo intermediate bam files removed.
echo All Done!
#+END_SRC
* MDA gatk_realign.sh pipeline
#+BEGIN_SRC sh
cd /scratch/bcb/hzhao1/scratch/clinsek_comparison
echo `date` start pre-processing for sample IPCT-CH-5446-Tumor-1290

echo `date` 
~/ipct/jdk1.7.0_51/bin/java -jar /scratch/bcb/hzhao1/tools/picard-tools-1.80/AddOrReplaceReadGroups.jar I=data/IPCT-CH-5446-Tumor-1290/bam/IPCT-CH-5446-Tumor-1290.rmdup.bam O=data/IPCT-CH-5446-Tumor-1290/bam/rg_added.IPCT-CH-5446-Tumor-1290.rmdup.bam RGPL=Illumina RGID=IPCT-CH-5446-Tumor-1290 RGSM=IPCT-CH-5446-Tumor-1290 RGLB=IPCT-CH-5446-Tumor-1290 RGPU=IPCT-CH-5446-Tumor-1290 SORT_ORDER=coordinate VALIDATION_STRINGENCY=LENIENT
~/tools/samtools-0.1.19/samtools index data/IPCT-CH-5446-Tumor-1290/bam/rg_added.IPCT-CH-5446-Tumor-1290.rmdup.bam
echo `date` AddOrReplaceReadGroups Done.

echo `date` start realignertargetCreator of IndelRealignments..
java -Xmx15g -jar GenomeAnalysisTK.jar -R /scratch/bcb/hzhao1/ipct/reference/hs37d5.fa -T RealignerTargetCreator -nt 4 -I data/IPCT-CH-5446-Tumor-1290/bam/rg_added.IPCT-CH-5446-Tumor-1290.rmdup.bam -o data/IPCT-CH-5446-Tumor-1290/bam/IPCT-CH-5446-Tumor-1290.intervals -known hs37_gatk_ref/1000G_phase1.indels.hs37.vcf -known hs37_gatk_ref/Mills_and_1000G_gold_standard.indels.hs37.vcf
echo `date` RealignerTargetCreator done.

echo `date` IndelRealignments..
java -Xmx15g -jar GenomeAnalysisTK.jar -R /scratch/bcb/hzhao1/ipct/reference/hs37d5.fa  -T IndelRealigner -I data/IPCT-CH-5446-Tumor-1290/bam/rg_added.IPCT-CH-5446-Tumor-1290.rmdup.bam -targetIntervals data/IPCT-CH-5446-Tumor-1290/bam/IPCT-CH-5446-Tumor-1290.intervals -o data/IPCT-CH-5446-Tumor-1290/bam/IPCT-CH-5446-Tumor-1290.IndelRealigned.bam -known hs37_gatk_ref/1000G_phase1.indels.hs37.vcf -known hs37_gatk_ref/Mills_and_1000G_gold_standard.indels.hs37.vcf
~/tools/samtools-0.1.19/samtools index data/IPCT-CH-5446-Tumor-1290/bam/IPCT-CH-5446-Tumor-1290.IndelRealigned.bam
echo `date` IndelRealigner and index Done. 

echo `date` BaseRecalibrator ..
java -Xmx15g -jar GenomeAnalysisTK.jar -R /scratch/bcb/hzhao1/ipct/reference/hs37d5.fa  -T BaseRecalibrator -I data/IPCT-CH-5446-Tumor-1290/bam/IPCT-CH-5446-Tumor-1290.IndelRealigned.bam -o data/IPCT-CH-5446-Tumor-1290/bam/IPCT-CH-5446-Tumor-1290.recal.table -knownSites hs37_gatk_ref/dbsnp138.hs37.vcf -knownSites hs37_gatk_ref/Mills_and_1000G_gold_standard.indels.hs37.vcf
echo `date` BaseRecalibrator done.

echo `date` Print reads..
java -Xmx15g -jar GenomeAnalysisTK.jar -R /scratch/bcb/hzhao1/ipct/reference/hs37d5.fa  -T PrintReads -I data/IPCT-CH-5446-Tumor-1290/bam/IPCT-CH-5446-Tumor-1290.IndelRealigned.bam -BQSR data/IPCT-CH-5446-Tumor-1290/bam/IPCT-CH-5446-Tumor-1290.recal.table -o data/IPCT-CH-5446-Tumor-1290/bam/IPCT-CH-5446-Tumor-1290.recaled.bam
echo `date` PrintReads done.

echo `date` index and flagstat the recalibrated bam
~/tools/samtools-0.1.19/samtools index data/IPCT-CH-5446-Tumor-1290/bam/IPCT-CH-5446-Tumor-1290.recaled.bam
~/tools/samtools-0.1.19/samtools flagstat data/IPCT-CH-5446-Tumor-1290/bam/IPCT-CH-5446-Tumor-1290.recaled.bam>data/IPCT-CH-5446-Tumor-1290/bam/IPCT-CH-5446-Tumor-1290.recaled.bam.flagstat
echo `date` Recalibrated BAM has been indexed.
echo Preprocessing of IPCT-CH-5446-Tumor-1290 is ALL DONE.
#+END_SRC
* perl utils usc

** fastqConstantSplit.pl
fastqConstantSplit.pl [#parts] [1.fq] [2.fq]

#+BEGIN_SRC
/home/uec-00/shared/production/software/perl_utils_usc/fastqConstantSplit.pl 1 /export/uec-gs1/laird/users/bberman/research/agilentRrbs/fromAgilent/KJ_SC112012_KJ80_I001_ATCACG_L005_R1.fastq /export/uec-gs1/laird/users/bberman/research/agilentRrbs/fromAgilent/KJ_SC112012_KJ80_I001_ATCACG_L005_R2.fastq
#+END_SRC

** wrap_bsmap.pl

wrap_bsmap.pl [result.bam] [ref.fa] [r1.fq] [r2.fq]

#+BEGIN_SRC 
/home/uec-00/shared/production/software/perl_utils_usc/wrap_bsmap.pl KJ_SC112012_KJ80_I001_ATCACG_L005_R1.fastq.NC_001416.fa.bam /home/uec-00/shared/production/genomes/lambdaphage/NC_001416.fa
/export/uec-gs1/laird/users/bberman/research/agilentRrbs/fromAgilent/KJ_SC112012_KJ80_I001_ATCACG_L005_R1.fastq /export/uec-gs1/laird/users/bberman/research/agilentRrbs/fromAgilent/KJ_SC112012_KJ80_I001_ATCACG_L005_R2.fastq
#+END_SRC

** orgcontamcheck.pl - contamination check
orgcontamcheck.pl [.fq] [noreads] [contam1.fa] [contam2.fa] [contam3.fa]...

#+BEGIN_SRC 
/home/uec-00/shared/production/software/perl_utils_usc/orgcontamcheck.pl /export/uec-gs1/laird/users/bberman/research/agilentRrbs/fromAgilent/KJ_SC112012_KJ80_I001_ATCACG_L005_R1.fastq 5000000  /home/uec-00/shared/production/genomes/encode_hg19_mf/female.hg19.fa  /home/uec-00/shared/production/genomes/sacCer1/sacCer1.fa  /home/uec-00/shared/production/genomes/phi-X174/phi_plus_SNPs.fa  /home/uec-00/shared/production/genomes/arabidopsis/tair8.pluscontam.fa  /home/uec-00/shared/production/genomes/mm9_unmasked/mm9_unmasked.fa  /home/uec-00/shared/production/genomes/Ecoli/EcoliIHE3034.fa  /home/uec-00/shared/production/genomes/rn4_unmasked/rn4.fa  /home/uec-00/shared/production/genomes/salmon/salmosalar.fa  /home/uec-00/shared/production/genomes/rRNA/rRNA.fa  /home/uec-00/shared/production/genomes/lambdaphage/NC_001416.fa 
#+END_SRC

** merge_bams.pl - Merge bam files

since FASTQ was aligned separately
this adds read groups using the Picard, mark duplicates and flagstat

merge_bams.pl [target.bam] [source1.bam] [source2.bam] ...

#+BEGIN_SRC
 /home/uec-00/shared/production/software/perl_utils_usc/merge_bams.pl ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam /export/uec-gs1/laird/shared/production/ga/flowcells/D295HACXX/results/D295HACXX/D295HACXX_7_NIC1254A76/ResultCount_D295HACXX_7_NIC1254A76.hg19_rCRSchrm.fa.mdups.bam /export/uec-gs1/laird/shared/production/ga/flowcells/C1LBAACXX/results/C1LBAACXX/C1LBAACXX_6_NIC1254A76/ResultCount_C1LBAACXX_6_NIC1254A76.hg19_rCRSchrm.fa.mdups.bam /export/uec-gs1/laird/shared/production/ga/flowcells/C27BLACXX/run2/results/C27BLACXX/C27BLACXX_6_NIC1254A76/ResultCount_C27BLACXX_6_NIC1254A76.hg19_rCRSchrm.fa.mdups.bam
#+END_SRC

or
#+BEGIN_SRC 
/home/uec-00/shared/production/software/perl_utils_usc/merge_bams.pl ResultCount_AGILENT_1_SC112012L005.NC_001416.fa.bam KJ_SC112012_KJ80_I001_ATCACG_L005_R1.fastq.NC_001416.fa.bam.rmuln.bam
#+END_SRC

** wrap_java_report.pl - merge QC metrics

wrap_java_report.pl MERGING_qcmetrics.csv bisulfiteQCMetrics.pl [dir]

#+BEGIN_SRC 
/home/uec-00/shared/production/software/perl_utils_usc/wrap_java_report.pl MERGING_qcmetrics.csv /home/uec-00/shared/production/software/perl_utils_usc/bisulfiteQCMetrics.pl /export/uec-gs1/laird/shared/production/ga/merges/2014-02-12_1437_merge_NIC1254A76/results/MERGING/MERGING_1_NIC1254A76
#+END_SRC

** wrap_lc_extrap.pl - coverage extrapolation

wrap_lc_extrap.pl [.bam] [.metric.txt] 

#+BEGIN_SRC 
/home/uec-00/shared/production/software/perl_utils_usc/wrap_lc_extrap.pl  ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam.CoverageProjection.metric.txt
#+END_SRC

** wrap_picard.pl
java -Xmx12g -jar picard.jar TMP_DIR=/export/uec-gs1/laird/shared/tmp
*** GC bias from Picard

wrap_picard.pl CollectBcBiasMerics.jar INPUT=[.bam] OUTPUT=[.CollectGcBiasMetrics.metric.txt]

#+BEGIN_SRC 
/home/uec-00/shared/production/software/perl_utils_usc/wrap_picard.pl CollectGcBiasMetrics.jar INPUT=ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam OUTPUT=ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam.CollectGcBiasMetrics.metric.txt VALIDATION_STRINGENCY=SILENT CHART_OUTPUT=chart REFERENCE_SEQUENCE=/home/uec-00/shared/production/genomes/hg19_rCRSchrm/hg19_rCRSchrm.fa
#+END_SRC

*** QualityScore Distribution

wrap_picard.pl QualityScoreDistribution.jar INPUT=[.bam] OUTPUT=[QualityScoreDistribution.metric.txt VALIDATION_STRINGENCY=SILENT CHART_OUTPUT=chart

#+BEGIN_SRC 
/home/uec-00/shared/production/software/perl_utils_usc/wrap_picard.pl QualityScoreDistribution.jar INPUT=ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam OUTPUT=ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam.QualityScoreDistribution.metric.txt VALIDATION_STRINGENCY=SILENT CHART_OUTPUT=chart
#+END_SRC

*** CollectAlignmentSummaryMetrics

wrap_picard.pl CollectAlignmentSummaryMetrics.jar INPUT=[.bam] OUTPUT=[.CollectAlignmentSummaryMetrics.metric.txt] IS_BISULFITE_SEQUENCED=true REFERENCE_SEQUENCE=[reference.fa]

#+BEGIN_SRC 
/home/uec-00/shared/production/software/perl_utils_usc/wrap_picard.pl CollectAlignmentSummaryMetrics.jar INPUT=ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.mdups.bam OUTPUT=ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.mdups.bam.CollectAlignmentSummaryMetrics.metric.txt IS_BISULFITE_SEQUENCED=true REFERENCE_SEQUENCE=/home/uec-00/shared/production/genomes/hg19_rCRSchrm/hg19_rCRSchrm.fa
#+END_SRC

*** CollectInsertSizeMetrics

wrap_picard.pl CollectInsertSizeMetrics.jar INPUT=[.bam] OUTPUT=[.CollectInsertSizeMetrics.metric.txt] VALIDATION_STRINGENCY=SILENT HISTOGRAM_FILE=chart

#+BEGIN_SRC
/home/uec-00/shared/production/software/perl_utils_usc/wrap_picard.pl CollectInsertSizeMetrics.jar INPUT=ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam OUTPUT=ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam.CollectInsertSizeMetrics.metric.txt VALIDATION_STRINGENCY=SILENT HISTOGRAM_FILE=chart
#+END_SRC

*** MeanQualityByCycle.jar

wrap_picard.pl MeanQualityByCycle.jar INPUT=[.bam] OUTPUT=[.MeanQualityByCycle.metric.txt] VALIDATION_STRINGENCY=SILENT CHART_OUTPUT=chart

#+BEGIN_SRC 
/home/uec-00/shared/production/software/perl_utils_usc/wrap_picard.pl MeanQualityByCycle.jar INPUT=ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam OUTPUT=ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam.MeanQualityByCycle.metric.txt VALIDATION_STRINGENCY=SILENT CHART_OUTPUT=chart
#+END_SRC

** bissnp_trinuc_sample.pl - MethLevelAverages by Yaping

NOTE: this depends on the bis-snp output:
*.raw.vcf.MethySummarizeList.txt

bissnp_trinuc_sampletrinuc_sample.pl [MethLevelAverages.metric.txt] [.bam] [reference.fa] [optional: chrM]

#+BEGIN_SRC
/home/uec-00/shared/production/software/perl_utils_usc/bissnp_trinuc_sample.pl ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam.MethLevelAverages.metric.txt ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam /home/uec-00/shared/production/genomes/hg19_rCRSchrm/hg19_rCRSchrm.fa
#+END_SRC

#+BEGIN_SRC
/home/uec-00/shared/production/software/perl_utils_usc/bissnp_trinuc_sample.pl ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam.chrM.MethLevelAverages.metric.txt ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam /home/uec-00/shared/production/genomes/hg19_rCRSchrm/hg19_rCRSchrm.fa chrM
#+END_SRC

** countnmer.pl
countnmer.pl [1.fq] [2.fq]

actually runs $JAVA edu.usc.epigenome.scripts.FastaToNmerCounts -nmer $i
#+BEGIN_SRC 
/home/uec-00/shared/production/software/perl_utils_usc/countnmer.pl nmerCount_AGILENT_s_1_3mers.csv KJ_SC112012_KJ80_I001_ATCACG_L005_R1.1.fastq KJ_SC112012_KJ80_I001_ATCACG_L005_R2.1.fastq 
#+END_SRC
** wrap_wigToBigWig.pl - wigToBigWig

wrap_wigToBigWig.pl [.wig] [.bw]

#+BEGIN_SRC 
/home/uec-00/shared/production/software/perl_utils_usc/wrap_wigToBigWig.pl ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam.winsize30dumpv.BinDepths.metric.wig ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam.winsize30dumpv.BinDepths.metric.wig.bw
#+END_SRC

#+BEGIN_SRC
/home/uec-00/shared/production/software/perl_utils_usc/wrap_wigToBigWig.pl ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam.winsize5000dumpv.BinDepths.metric.wig ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam.winsize5000dumpv.BinDepths.metric.wig.bw
#+END_SRC
** trackversions.pl - track versions

trackversions.pl [.bam] [.ApplicationStackMetrics.metric.txt]

#+BEGIN_SRC 
/home/uec-00/shared/production/software/perl_utils_usc/trackversions.pl ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam.ApplicationStackMetrics.metric.txt
#+END_SRC

** cleanPipelineParts.pl - clean pipeline parts

cleanPipelineParts.pl [dir]

#+BEGIN_SRC 
/home/uec-00/shared/production/software/perl_utils_usc/cleanPipelineParts.pl /export/uec-gs1/laird/shared/production/ga/merges/2014-02-12_1437_merge_NIC1254A76/results/MERGING/MERGING_1_NIC1254A76
#+END_SRC

** wrap_rnaseq_metrics.pl

wrap_rnaseq_metrics.pl [.bam] [.fa] [.CollectRnaSeqMetrics.metric.txt]

#+BEGIN_SRC
/home/uec-00/shared/production/software/perl_utils_usc/wrap_rnaseq_metrics.pl ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.mdups.bam /home/uec-00/shared/production/genomes/hg19_rCRSchrm/hg19_rCRSchrm.fa ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.mdups.bam.CollectRnaSeqMetrics.metric.txt
#+END_SRC

** bamToElementEnrichment.pl

bamToElementEnrichment.pl [.bam] [.CPGvsRandomCov.metric.txt]

#+BEGIN_SRC
/home/uec-00/shared/production/software/perl_utils_usc/bamToElementEnrichment.pl  ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam /home/rcf-40/bberman/tumor/genomic-data-misc/CGIs/Takai_Jones_from_Fei_122007.fixed.PROMOTERONLY.oriented.hg19.bed ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam.CPGvsRandomCov.metric.txt
#+END_SRC

** solexaFilterContams.pl
* uecgatk.pl
** bin depths

uecgatk.pl -T BinDepths -R [reference.fa] -I [.bam] -o [.wig]

#+BEGIN_SRC
/home/uec-00/shared/production/software/uecgatk/default/uecgatk.pl  -T BinDepths -R /home/uec-00/shared/production/genomes/hg19_rCRSchrm/hg19_rCRSchrm.fa -I ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam -o ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam.winsize30dumpv.BinDepths.metric.wig -winsize 30 -dumpv
#+END_SRC

#+BEGIN_SRC
/home/uec-00/shared/production/software/uecgatk/default/uecgatk.pl  -T BinDepths -R /home/uec-00/shared/production/genomes/hg19_rCRSchrm/hg19_rCRSchrm.fa -I ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam -o ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam.p5000000winsize50000dumpv.BinDepths.metric.wig -p 5000000 -winsize 5000 -dumpv
#+END_SRC

#+BEGIN_SRC
/home/uec-00/shared/production/software/uecgatk/default/uecgatk.pl  -T BinDepths -R /home/uec-00/shared/production/genomes/hg19_rCRSchrm/hg19_rCRSchrm.fa -I ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam -o ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam.winsize5000dumpv.BinDepths.metric.wig -winsize 5000 -dumpv
#+END_SRC

** DownsampleDups - downsample duplicates

uecgatk.pl -T DownsampleDups -R [reference.fa] -I [.bam] -o [DownsampleDups.metric.txt]

#+BEGIN_SRC 
/home/uec-00/shared/production/software/uecgatk/default/uecgatk.pl  -T DownsampleDups -R /home/uec-00/shared/production/genomes/hg19_rCRSchrm/hg19_rCRSchrm.fa -I ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam -o ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam.p5000000trials100nt8.DownsampleDups.metric.txt -p 5000000 -trials 100 -nt 8
#+END_SRC

** Read length metric

uecgatk.pl -T ReadLength -R [.fa] -I [.bam] -o [ReadLength.metric.txt]

#+BEGIN_SRC 
/home/uec-00/shared/production/software/uecgatk/default/uecgatk.pl  -T ReadLength -R /home/uec-00/shared/production/genomes/hg19_rCRSchrm/hg19_rCRSchrm.fa -I ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam -o ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam..ReadLength.metric.txt
#+END_SRC

** MethLevelAverages

uecgatk.pl -T MethLevelAverages -R [reference.fa] -I [.bam] -o [.cph.MethLevelAverages.metric.txt] -cph

specify chromosomes: -L chrM:1234-12345

#+BEGIN_SRC 
/home/uec-00/shared/production/software/uecgatk/default/uecgatk.pl  -T MethLevelAverages -R /home/uec-00/shared/production/genomes/hg19_rCRSchrm/hg19_rCRSchrm.fa -I ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam -o ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam.cph.MethLevelAverages.metric.txt -cph
#+END_SRC

** FlagStat

uecgatk.pl -T FlagStat -R [reference.fa] -I [.bam] -o [.FlagStat.metric.txt]

#+BEGIN_SRC 
/home/uec-00/shared/production/software/uecgatk2/default/uecgatk.pl  -T FlagStat -R /home/uec-00/shared/production/genomes/hg19_rCRSchrm/hg19_rCRSchrm.fa -I ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.mdups.bam -o ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.mdups.bam.rfNumberHitsEqualsOne.FlagStat.metric.txt -rf NumberHitsEqualsOne
#+END_SRC

** InvertedReadPairDups

uecgatk.pl -T InvertedReadPairDups -R [reference.fa] -I [.bam] -o [.InvertedReadPairDups.metric.txt]

#+BEGIN_SRC 
/home/uec-00/shared/production/software/uecgatk/default/uecgatk.pl  -T InvertedReadPairDups -R /home/uec-00/shared/production/genomes/hg19_rCRSchrm/hg19_rCRSchrm.fa -I ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam -o ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam..InvertedReadPairDups.metric.txt
#+END_SRC

* bissnp pipeline

bissnp_pipeline.pl [.bam] [reference.fa]

#+BEGIN_SRC 
/home/uec-00/shared/production/software/bissnp/bissnp_pipeline.pl ResultCount_MERGING_1_NIC1254A76.hg19_rCRSchrm.fa.bam /home/uec-00/shared/production/genomes/hg19_rCRSchrm/hg19_rCRSchrm.fa
#+END_SRC

** bam_indel_realign

java -jar bissnp.jar -R 

** bam_mdups

** bam_base_recalibration

** bissnp

** vcf_sort

** vcf_filter

** vcf2bed6plus2

** vcf2bed6plus2_filter

** vcf2tdf

* GATK
** CountReads - count reads from bam file
java -jar $GATK -R [reference.fasta] -I [.bam]
** CountLoci - count how many bases have been covered by at least one or more reads
java -jar $GATK -T CountLoci -R GATK-3.3.0/resources/exampleFASTA.fasta -I GATK-3.3.0/resources/exampleBAM.bam -o output.txt
* Picard
** AddOrReplaceReadGroups

#+BEGIN_SRC 
$JAVABIN -Xmx4g -jar ~/tools/picard/picard-tools-1.128/picard.jar AddOrReplaceReadGroups INPUT=output/NIC1254A45.bam OUTPUT=output/NIC1254A45_RG.bam RGLB=MERGING RGPL=illumina RGPU=1 RGSM=NIC1254A45
#+END_SRC

or old picard
#+BEGIN_SRC 
$JAVA -Xmx4g -jar $PICARD/AddOrReplaceReadGroups.jar CREATE_INDEX=true VALIDATION_STRINGENCY=SILENT SORT_ORDER=coordinate
MAX_RECORDS_IN_RAM=1000000 INPUT='$bamIn' OUTPUT='with_rg_$bamOut' RGID='$flowcell\.$lane' RGLB='$lib' RGPL='illumina Hiseq'
RGPU='$flowcell\.$lane' RGSM='$lib' RGCN='USC EPIGENOME CENTER' RGDS='from file $bamIn on $date'"
#+END_SRC

** MergeSamFiles

$JAVA -Xmx4g -jar picard.jar MergeSamFiles INPUT=[in1.bam] INPUT=[in2.bam] ... OUTPUT=[out.bam]

or old picard
#+BEGIN_SRC
$JAVA -Xmx4g -jar $PICARD/MergeSamFiles.jar $cmd
#+END_SRC

** MarkDuplicate

#+BEGIN_SRC 
$JAVA -Xms7g -Xmx7g -jar $PICARD/MarkDuplicates.jar CREATE_INDEX=true VALIDATION_STRINGENCY=SILENT METRICS_FILE=dupmets.txt
READ_NAME_REGEX=null INPUT=$output OUTPUT=$outputdups
#+END_SRC

